enhancements/network/cloud-egress-ip.md

```diff
@@ -0,0 +1,520 @@
+provided and always use a healthy node (thus not requiring the network plugin to
+perform an egress IP re-assignment every time a route goes down).
+
+Given the above, it is however important that the network plugin is conveyed the
+cloud's information concerning the nodes' subnet range. This is especially
+evident on GCP, as without it: the network plugin won't be able to perform any
+assignments at all, given the nodes' default `/32` subnet.
+
+##### Inter-component communication
+
+Multiple pieces of information needs to be communicated between the network
+plugins and this cloud component:
+
+1. The cloud's node subnet information.
+2. The egress IP assignment to be performed by the cloud component
+3. The status of the egress IP assignment performed by the cloud component
+
+Two models have been decided for this: a CRD based approach and a node
+annotated approach.
+
+**annotated approach**
+
+As to satisfy condition 1. the cloud component will annotate all cluster nodes
+with the nodes' cloud subnet as retrieved from the cloud provider's API when it
+is initialized. This annotation will have the following specification:
+
+`network.openshift.io/node-cloud-ifaddr: {"ipv4": "$IPv4_ADDRESS/$IPv4_SUBNET_MASK", "ipv6": "$IPv6_ADDRESS/$IPv6_SUBNET_MASK"}`
+
+This will in turn be read by the network plugin and used during the egress IP
+assignment, superseding any host level subnet information already parsed by the
+network plugin.
+
+**CRD**
+
+The cloud component will define a dedicated CRD that the network plugin will use
+to issue cloud egress IP requests. The network plugin will in-turn finalize its
+setup and set its final egress IP status pending the CR's status indicating a
+successful/unsuccessful cloud assignment. For example: if the cloud assignment
+should be unsuccessful; it will be up to the network plugin to notify the user
+of this, by setting its own status field for the egress IP assignment
+accordingly.
+
+The CRD will have `scope: Cluster` (since egress IP is cluster-scoped and node
+private IP assignment have no relation to namespaces) and have the following
+format:
+
+```yaml
+apiVersion: network.openshift.io/v1
+kind: CloudPrivateIPConfig
+metadata:
+  name: cloud-egressip
+spec:
+  items:
+  - node: nodeX
+    egressIP: 192.168.126.11
+  - node: nodeY
+    egressIP: 192.168.126.102
+status:
+  items:
+  - node: nodeX
+    egressIP: 192.168.126.11
+  - node: nodeY
+    egressIP: 192.168.126.102
+```
+
+where `nodeX`/`nodeY` will be cluster node names as retrieved by
+`nodes[*].metadata.name`. The referential order between `.spec.items` and
+`.status.items` will not be strict, meaning `.spec.items[0]` will not
+necessarily correspond to `.spec.items[0]`.
+
+In case an assignment is unsuccessful: the `.status.items[?]` corresponding
+to that `.spec.items[?]` will not be set. This is done as to explicitly
+notify the network plugin of which (if not all) requests have failed by still
+issuing an UPDATE on the object.
```

Comment:
This is not a user facing CRD. A user of OpenShift cares about the functionality egress IP, and only the network plugins will use this feature to make sure it functions in the cloud. The user of this, is thus the networking team (or Red Hat internal), and we should check the logs of this component is there's an unsuccessful assignment.

Reply:
> This is not a user facing CRD.

It isn't obvious to me that this is the intention currently, this is probably my lack of understanding of the space.

Also, bit tongue in cheek but, can a CRD ever really be "not user facing"? What stops a user who has done an IPI installation from creating/reading/modifying this kind of resource? What happens when this isn't working correctly, someone will want to work out why it's broken, that someone may be a RH SRE, a customer or even a developer, would it not be easier for them to work out what broke if we put it on the status than making them trawl logs? This is completely up to you if you class it as internal but I'd encourage you to think about this.

Reply:
Nothing can stop a cluster admin from creating such CRs..but the idea is (and again, I have not stated this in the enhancement proposal, and I will do so now) to have specific service accounts (the ones associated with the network plugins) enforce that this creation is only done by them.  

> What happens when this isn't working correctly, someone will want to work out why it's broken

Concerning the status of the assignment, and why any such assignment might have failed: please note that the network plugins don't explicitly have such a status field. So even if this component did: the egress IP feature - which when it comes down to it, is the only really user facing feature - does not. We could add a status field to each item, but I don't see how any status could cover all possible failure modes for this and give the user a clear indicative trace of what went wrong. I could have something along the lines of:

```
apiVersion: network.openshift.io/v1
kind: CloudPrivateIPConfig
metadata:
  name: cloud-egressip
spec:
  items:
  - node: nodeX
    egressIP: 192.168.126.11
  - node: nodeY
    egressIP: 192.168.126.102
status:
  items:
  - node: nodeX
    egressIP: 192.168.126.11
    status: SUCCESS
  - node: nodeY
    egressIP: 192.168.126.102
    status: ERROR
```
but I don't see what benefit of that vs:

```
apiVersion: network.openshift.io/v1
kind: CloudPrivateIPConfig
metadata:
  name: cloud-egressip
spec:
  items:
  - node: nodeX
    egressIP: 192.168.126.11
  - node: nodeY
    egressIP: 192.168.126.102
status:
  items:
  - node: nodeX
    egressIP: 192.168.126.11
```
is, and what it would give anyone. One is explicitly stated as in error, whereas the other is implicitly stated. But you still need to look in logs to understand why in both cases. 

That's my opinion on the matter, but if the consensus gravitates towards a status field, I could change that.